### PR TITLE
[SHARE-632][Feature] MODS transformer

### PR DESCRIPTION
**Pull request by aaxelb**
Add MODS transformer to support TDAR and others.

Add a MODS source config for every OAI source that lists it as a supported metadata prefix, and disable the old `oai_dc` source config.

Continues work in #597

### CODE REVIEW

**Comment by leb2dg** [comment]
@aaxelb, there is a failing test

**Comment by aaxelb** [comment]
@leb2dg I couldn't reproduce that failing test, and neither could travis. Seems like stevedore nondeterministically clashes with our mocks in a way I don't understand...

File: share/transform/chain/links.py [hunk 1]
```diff
@@ -484,6 +489,9 @@ def __init__(self, parser):
         super().__init__()
 
     def execute(self, obj):
+        # callable will return True for classes as well as functions
+        if isinstance(self._parser, types.FunctionType):
+            self._parser = self._parser(obj)
```
Comment (leb2dg) [inline share/transform/chain/links.py]: After team discussion, this seemed to cause some issues with the agent_parser. Will send back to additional development until updated.

File: share/harvesters/oai.py [hunk 2]
```diff
@@ -103,3 +103,12 @@ def fetch_by_id(self, provider_id):
         url.args['metadataPrefix'] = self.metadata_prefix
         url.args['identifier'] = provider_id
         return etree.tostring(self.fetch_page(url)[0][0], encoding=str)
+
+    def metadata_formats(self):
```
Comment (leb2dg) [inline share/harvesters/oai.py]: Consider moving to another PR since this isn't used my MODS (per our discussion).

**Comment by leb2dg** [comment]
When running org.bhl.mods:
```
[2017-03-24 20:17:21,143: ERROR/MainProcess] Task share.tasks.NormalizerTask[535b4b96-f285-4926-b235-ab5e4508fdba] raised unexpected: KeyError('#text',)
Traceback (most recent call last):
  File "/Users/laurenbarker/.virtualenvs/share/lib/python3.5/site-packages/celery/app/trace.py", line 240, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/Users/laurenbarker/.virtualenvs/share/lib/python3.5/site-packages/celery/app/trace.py", line 438, in __protected_call__
    return self.run(*args, **kwargs)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/tasks.py", line 71, in run
    self.do_run(*self.args, **self.kwargs)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/tasks.py", line 335, in do_run
    raise self.retry(countdown=10, exc=e)
  File "/Users/laurenbarker/.virtualenvs/share/lib/python3.5/site-packages/celery/app/task.py", line 676, in retry
    maybe_reraise()
  File "/Users/laurenbarker/.virtualenvs/share/lib/python3.5/site-packages/celery/utils/__init__.py", line 248, in maybe_reraise
    reraise(exc_info[0], exc_info[1], exc_info[2])
  File "/Users/laurenbarker/.virtualenvs/share/lib/python3.5/site-packages/celery/five.py", line 91, in reraise
    raise value
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/tasks.py", line 316, in do_run
    graph = transformer.transform(raw)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/base.py", line 22, in transform
    jsonld, root_ref = self.do_transform(datum)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transformers/mods.py", line 484, in do_transform
    return super().do_transform(data)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/transformer.py", line 50, in do_transform
    root_ref = parser(unwrapped).parse()
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/parsers.py", line 75, in parse
    return self._do_parse()
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/parsers.py", line 96, in _do_parse
    value = chain.run(self.context)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 220, in run
    return self.execute(obj)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 232, in execute
    return reduce(lambda acc, cur: cur.run(acc), self.chain()[1:], obj)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 232, in <lambda>
    return reduce(lambda acc, cur: cur.run(acc), self.chain()[1:], obj)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 220, in run
    return self.execute(obj)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 338, in execute
    for chain in self._chains
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 338, in <listcomp>
    for chain in self._chains
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 220, in run
    return self.execute(obj)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 232, in execute
    return reduce(lambda acc, cur: cur.run(acc), self.chain()[1:], obj)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 232, in <lambda>
    return reduce(lambda acc, cur: cur.run(acc), self.chain()[1:], obj)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 220, in run
    return self.execute(obj)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 338, in execute
    for chain in self._chains
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 338, in <listcomp>
    for chain in self._chains
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 220, in run
    return self.execute(obj)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 232, in execute
    return reduce(lambda acc, cur: cur.run(acc), self.chain()[1:], obj)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 232, in <lambda>
    return reduce(lambda acc, cur: cur.run(acc), self.chain()[1:], obj)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 220, in run
    return self.execute(obj)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 338, in execute
    for chain in self._chains
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 338, in <listcomp>
    for chain in self._chains
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 220, in run
    return self.execute(obj)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 232, in execute
    return reduce(lambda acc, cur: cur.run(acc), self.chain()[1:], obj)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 232, in <lambda>
    return reduce(lambda acc, cur: cur.run(acc), self.chain()[1:], obj)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 220, in run
    return self.execute(obj)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 376, in execute
    return [self.__anchor.run(sub) for sub in obj]
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 376, in <listcomp>
    return [self.__anchor.run(sub) for sub in obj]
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 220, in run
    return self.execute(obj)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 232, in execute
    return reduce(lambda acc, cur: cur.run(acc), self.chain()[1:], obj)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 232, in <lambda>
    return reduce(lambda acc, cur: cur.run(acc), self.chain()[1:], obj)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 220, in run
    return self.execute(obj)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transform/chain/links.py", line 508, in execute
    return self._function_name(obj, *self._args, **self._kwargs)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/transformers/mods.py", line 17, in force_text
    return data['#text']
KeyError: '#text'
```

**Comment by aaxelb** [comment]
@leb2dg fixed! 🔧

**Comment by leb2dg** [comment]
Looks like all contributors aren't being pulled out. Some works from `be.ghent.mods` have missing contributors.

**Comment by leb2dg** [comment]
Null titles in related works are getting created with `title=.` (`org.tdar.mods`)

**Comment by aaxelb** [comment]
Thanks @leb2dg! It was missing all non-creator/funder/publisher/host contributors because stupid bug, and got confused by empty title parts. Both should be fixed, now.

**Comment by chrisseto** [comment]
@aaxelb could we go ahead and split non-tdar source configs into another PR?

File: share/transformers/mods.py [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+        name_parts = get_list(name, 'mods:namePart')
+        return ' '.join([force_text(n) for n in name_parts])
+
+    def get_agent_schema(self, obj):
+        name_type = obj.get('@type')
+        if name_type == 'personal':
+            return 'person'
+        if name_type == 'conference':
+            return 'organization'
+        # if name_type == 'family':
```
Comment (chrisseto) [inline share/transformers/mods.py]: Should we make a ticket for this?

File: share/transformers/mods.py [hunk 4]
```diff
@@ -0,0 +1,491 @@
...
+        tools.OneOf(
+            tools.RunPython(force_text, ctx['mods:genre']),
+            tools.Static(None)
+        )
+    )
+
+    title = tools.RunPython('join_title_info', ctx)
+
+    # Abstracts have the optional attribute "shareable". Don't bother checking for it, because
+    # abstracts that are not shareable should not have been shared with SHARE.
```
Comment (chrisseto) [inline share/transformers/mods.py]: 👍

File: share/transformers/mods.py [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+    )
+
+    tags = tools.Map(
+        tools.Delegate(MODSThroughTags),
+        tools.Concat(
+            tools.Map(
+                tools.RunPython('tokenize'),
+                tools.Map(
+                    tools.RunPython(force_text),
+                    tools.Concat(
```
Comment (chrisseto) [inline share/transformers/mods.py]: I don't think you need `Concat` when calling `Map`? It should implicitly call it for you.

**Comment by chrisseto** [comment]
Going to do one last check with a local run, otherwise looks gtg

**Comment by aaxelb** [comment]
@chrisseto Moved all the non-tdar configs to #606

**Comment by chrisseto** [comment]
<3

File: share/transformers/mods.py (hunk 5 above)
Comment (aaxelb) [inline share/transformers/mods.py]: Yep, you're right

File: share/transformers/mods.py (hunk 3 above)
Comment (aaxelb) [inline share/transformers/mods.py]: [Done](https://openscience.atlassian.net/browse/SHARE-718)

**Comment by chrisseto** [comment]
@aaxelb I'm not seeing an identifiers other than the OAI id.
Shouldn't `https://core.tdar.org/document/401234/printed-draft-map-salamonie-reservoir-maps-nd` and the DOI get pulled out of this one?
```xml
            <ns6:identifier type="uri">https://core.tdar.org/document/401234/printed-draft-map-salamonie-reservoir-maps-nd</ns6:identifier>
            <ns6:identifier
                type="tdarId">401234</ns6:identifier>
                <ns6:identifier type="doi">doi:10.6067/XCV8X068SJ</ns6:identifier>
                <ns6:originInfo>
                    <ns6:publisher>US Army Corps of Engineers Mandatory Center of Expertise for the Curation and Management of Archaeological
                        Collections, St. Louis District</ns6:publisher>
                    <ns6:dateCreated>2012</ns6:dateCreated>
                </ns6:originInfo>
                <ns6:language>
                    <ns6:languageTerm type="text">en</ns6:languageTerm>
                </ns6:language>
                <ns6:typeOfResource>text</ns6:typeOfResource>
                <ns6:abstract>This resource contains a four-piece printed draft map of the Salamonie Reservoir, Indiana, printed in February
                    1981. This map is part of the Salamonie Reservoir Maps N.D collection that consists of oversized maps
                    of the Salamonie Reservoir within the Huntington and Wabash Counties, Indiana. The collection was received
                    without any association to a particular investigation. The date of this image was not recorded. This
                    is likely due to the documents’ large size and cartographic contents.</ns6:abstract>
                <ns6:identifier type="doi"
                />
```

**Comment by aaxelb** [comment]
@chrisseto Oh whoops, sorry, it's fixed now. At one point I changed `force_text` to always return a string, including joining lists together, but didn't notice it was then squishing all the identifiers together and then finding multiple IRIs and failing...
